### PR TITLE
Add basic WebviewView-based query result panel

### PR DESCRIPTION
**Pull request by caohai**
This PR implements the basic features of the new react based query result pane.
This PR also makes the `WebviewRpc` class a singleton to ensure responses always get sent back to the same client where the request was created.
image
image

Files under `src/reactviews/pages/QueryResult/table` are all directly ported from ADS (with some build/typing fixes).

To test this feature out, follow the following steps:
1. checkout `dev/query-result`, build and start the extension
2. Open VSCode setting file `Setttings.json`, add `"mssql.enableNewQueryResultFeature": true`, this flag is for dev purpose and will later be merged into the preview feature flag.
3. Reload the window
4. Set `"mssql.enableNewQueryResultFeature"` to `false` will go back to the old angular implementation

Missing features will be implemented later in smaller size PRs, some will need more design discussions.
1. Multiple result set support: only 1 batch is supported for now
2. Grid:
     - Plugin features: sorting/filtering/saving
     - Auto-resizing
     - Styling
3. Results:
    - Styling
    - Execution time message
    - Link handling
 4. Snapshot feature

### CODE REVIEW

File: package.json
Comment (aasimkhan30) [inline package.json:1]: Probably add config option to enable grid over here?

File: src/controllers/reactWebviewViewController.ts
Comment (aasimkhan30) [inline src/controllers/reactWebviewViewController.ts:1]: This file was renamed to reactWebviewController. Maybe a merge issue?

File: src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts:1]: is this file needed?

File: package.json
Comment (caohai) [reply package.json:1]: I'm currently hiding the flag for dev purpose but will eventually remove this flag and merge the new grid feature into the preview flag we have.

File: src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts
Comment (caohai) [reply src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts:1]: @cssuh Do you think we'll need this file later?

File: src/reactviews/pages/QueryResult/slickgrid.tsx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                // If the string only contains whitespaces, it will be treated as empty string to make the filtering easier.
+                // Note: this is the display string and does not impact the export/copy features.
+                return data.displayValue.trim() === '' ? '' : data.displayValue;
+            },
+            {
+                inMemoryDataProcessing: true,
+            },
+            undefined,
+            undefined);
+        var table = new Table(div, defaultTableStyles, { dataProvider: dataProvider, columns: columns }, tableOptions);
```
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/slickgrid.tsx:125]: @caohai , can you please make there are no memory leaks and we are clearing the old div and table before creating a new one here?

File: src/reactviews/pages/QueryResult/slickgrid.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+        collection.setCollectionChangedCallback((startIndex, count) => {
+            let refreshedRows = range(startIndex, startIndex + count);
+            table.invalidateRows(refreshedRows, true);
+        });
+        table.updateRowCount();
+
+        let grid = document.body.appendChild(div);
+        const elm = document.getElementById('grid')!;
+        document.body.removeChild(grid);
```
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/slickgrid.tsx:135]: I see that we are clearing out the old grid. What about old Table object? Is that getting disposed properly?

File: src/reactviews/pages/QueryResult/slickgrid.tsx
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/slickgrid.tsx:1]: Please profile this component in `dev tools` to make sure we don't have memory leaks.

Comment (caohai) [reply src/reactviews/pages/QueryResult/slickgrid.tsx:1]: I'll sync with @cssuh and probably reach out to you to do this.

**Comment by aasimkhan30** [comment]
@caohai, should we change the background color for result cells to black? Maybe the grid is not picking up the CSS styling but the background colors for header cells and result cells look pretty similar. 

image

File: src/controllers/reactWebviewViewController.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * ReactWebviewViewController is a class that manages a vscode.WebviewView and provides
+ * a way to communicate with it. It provides a way to register request handlers and reducers
+ * that can be called from the webview. It also provides a way to post notifications to the webview.
+ * @template State The type of the state object that the webview will use
+ * @template Reducers The type of the reducers that the webview will use
+ */
+export class ReactWebviewViewController<State, Reducers> extends ReactWebviewBaseController<State, Reducers> implements vscode.WebviewViewProvider{
+	private _webviewView: vscode.WebviewView;
+
```
Comment (aasimkhan30) [inline src/controllers/reactWebviewViewController.ts:18]: NIT: whitespace

**Comment by caohai** [comment]
> @caohai, should we change the background color for result cells to black? Maybe the grid is not picking up the CSS styling but the background colors for header cells and result cells look pretty similar.
> 
> image

@aasimkhan30 I don't think styling is properly setup right now (I've listed it as part of the missing features), @cssuh will be working on correctly setting up the grid styling with the ported css files.

File: src/reactviews/pages/QueryResult/queryResultPane.tsx [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+			<TabList
+				size='medium'
+				selectedValue={metadata.tabStates!.resultPaneTab}
+				onTabSelect={(_event, data) => {
+					state?.provider.setResultTab(data.value as qr.QueryResultPaneTabs);
+				}}
+				className={classes.queryResultPaneTabs}
+			>
+				{metadata.resultSetSummary && <Tab value={qr.QueryResultPaneTabs.Results} key={qr.QueryResultPaneTabs.Results}>
+					Results
```
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/queryResultPane.tsx:141]: Please localize the strings in a follow up PR.

File: src/reactviews/pages/QueryResult/queryResultPane.tsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+			case webDarkTheme:
+				return 'vs-dark';
+			case teamsHighContrastTheme:
+				return 'hc-black';
+			default:
+				return 'light';
+		}
+	};
+
+	getVscodeTheme;
```
Comment (aasimkhan30) [inline src/reactviews/pages/QueryResult/queryResultPane.tsx:86]: is this needed?

Comment (caohai) [reply src/reactviews/pages/QueryResult/queryResultPane.tsx:86]: no it's not, added it for debugging, will remove

**Review by aasimkhan30 — APPROVED** [review]

**Comment by aasimkhan30** [comment]
Just did some basic profiling on the branch. The memory and node count seem to remain constant.

File: src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts
Comment (cssuh) [reply src/reactviews/pages/QueryResult/table/plugins/mousewheelTableScroll.plugin.ts:1]: I'll remove this for now

File: src/reactviews/pages/QueryResult/queryResultPane.tsx (hunk 5 above)
Comment (caohai) [reply src/reactviews/pages/QueryResult/queryResultPane.tsx]: removed

File: src/reactviews/pages/QueryResult/queryResultPane.tsx (hunk 4 above)
Comment (caohai) [reply src/reactviews/pages/QueryResult/queryResultPane.tsx]: updated

File: src/controllers/reactWebviewViewController.ts (hunk 3 above)
Comment (caohai) [reply src/controllers/reactWebviewViewController.ts:18]: removed
